Gadget render optimization

danieldresser-ie opened the pull request:
Here's the stuff I've been working for improving Gadget render performance.  Definitely look through it carefully, but the overall approach is looking quite good - it's doing a lot more than reducing the overhead from the Focus PR.

Oh, I should throw out the first commit that dumps the frametime before we merge, but it is pretty useful when evaluating this.

Here are the main metrics I've been considering, measured by eyeballing average milliseconds per frame from the console output, testing in a scene with 4000 group nodes connected in a chain.  The places where "16 milliseconds" shows up indicate that we're actually idling wait for the next refresh, the actual time spent rendering is actually a bit lower.

|      |      Original      |  This PR |
|----------|:-------------:|------:|
| Panning ( Zoomed out ) | 81 | 49 |
|  Panning ( Zoomed in ) | 81 | 16 |
| Moving Node ( Zoomed out ) | 120 | 85  |
| Moving Node ( Zoomed in ) | 120 | 51 |
| Dragging Noodle ( Zoomed out ) | 180 | 66 |
| Dragging Noodle ( Zoomed in ) | 180 | 16 |
| Just Moving Cursor ( running gadgetsAt ) | 107 | 7 |

johnhaddon reviewed (commented):
Thanks for taking the initiative on this Daniel - the speedups really do feel pretty significant, and I think this will be well received.

Comments inline as usual, but I've also uncovered a few bugs in testing :

- Make a Sphere node. Drag a connection from the `out` nodule into the edge of the viewport to initialise the auto-pan mechanism. Once the connection start-point is auto-scrolled off the edge of the screen, the connection stops drawing.
- Make a Camera node. Select the camera in the 3d viewer so that its frustum is shown. Move around so that the body of the camera is offscreen but the frustum is still onscreen. The frustum will sometimes disappear (depending on whether or not the body can be clipped).
- Nodule labels are no longer being displayed on hover.

I'm a bit concerned that the first one at least might imply that we need a different bound for layout than we do for rendering? I'm hoping we can sort something out without needing too much additional machinery...

danieldresser-ie commented on src/GafferUI/AnnotationsGadget.cpp:
```diff
@@ -249,6 +249,11 @@ void AnnotationsGadget::doRenderLayer( Layer layer, const Style *style ) const
 
+int AnnotationsGadget::hasLayer() const
+{
+	return (int)GraphLayer::Overlay;
```
Sounds reasonable, but not actually possible to do for an enum, since conversions have to be declared inside the class?  At least, that's what I'm getting from here: https://groups.google.com/a/isocpp.org/g/std-proposals/c/eiN1Xqrt59Y

johnhaddon replied:
OK, explicit cast it is then.

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -133,7 +133,7 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
 		void setVisible( bool visible );
 		/// Returns the visibility status for this Gadget.
-		bool getVisible() const;
+		inline bool getVisible() const { return m_visible; }
```
Fixed in place.

danieldresser-ie commented on src/GafferUI/Gadget.cpp:
```diff
@@ -326,6 +326,10 @@ void Gadget::dirty( DirtyType dirtyType )
 				viewportGadget->renderRequestSignal()( viewportGadget );
+				if( dirtyType == DirtyType::Layout )
+				{
+					viewportGadget->dirtyLayout();
```
Added a dummy function to this commit so that it compiles in this state.

danieldresser-ie commented on include/GafferUI/ViewportGadget.h:
```diff
@@ -246,8 +246,23 @@ class GAFFERUI_API ViewportGadget : public Gadget
+		/// We need to be notified when the layout is dirtied so that we can rebuild the render items list.
+		/// Because we can't override Gadget::dirty(), I've added a hack where Gadget notifies ViewportGadget
+		/// specifically
+		void dirtyLayout();
```
Fixed in beedca6

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1035,72 +1044,93 @@ void ViewportGadget::render() const
+			boundDefault ? g_infiniteBox : Imath::transform( bound, transform )
+		} );
+	}
+	const M44f &inheritTransform = renderItems.back().transform;
```
Simplified logic in c2e2b48 avoids this issue.

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1035,72 +1044,93 @@ void ViewportGadget::render() const
+		renderItems.push_back( {
+			gadget, transform,
+			boundDefault ? g_infiniteBox : Imath::transform( bound, transform )
+		} );
```
I did see a small ( percent or two ) performance improvement in some cases, so I kept the test against identity ... with the new logic in c2e2b48 I feel like it's probably a lot less annoying anyway.

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -88,13 +88,12 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
-			Main = 0,
-			MidFront = 1,
-			Front = 2,
+			None = -1,
```
Addressed in place in 6eff5b5.  I think I'm recalling correctly that when we talked it through, we decided not to worry much about headroom because we can adjust this while keeping source compatibility?  Or should I still bump it a bit?

danieldresser-ie commented on include/GafferUI/Gadget.h:
```diff
@@ -290,10 +289,11 @@ class GAFFERUI_API Gadget : public Gaffer::GraphComponent
+		/// Returns a bitmask built from the flags in the Layer enum
+		/// Any subclass which implements doRenderLayer must also implement hasLayer
+		/// to indicate which layers doRender should be called for
+		virtual int hasLayer() const;
```
Addressed in place in 6eff5b5

danieldresser-ie commented on src/GafferSceneUI/UVView.cpp:
```diff
@@ -474,6 +474,11 @@ class GridGadget : public GafferUI::Gadget
 
+		int hasLayer() const override
+		{
+			return (int)Layer::Main | (int)Layer::MidBack | (int)Layer::Front;
```
Addressed in place in 6eff5b5.  I've added an `operator|` for both two Layers, and for unsigned and Layer ( allowing chains of layers like `Main | MidBack | Front`

danieldresser-ie commented on src/GafferUI/StandardNodule.cpp:
```diff
@@ -226,6 +204,16 @@ void StandardNodule::doRenderLayer( Layer layer, const Style *style ) const
+		(int)GraphLayer::Connections |
+		(int)GraphLayer::Nodes |
+		(int)GraphLayer::Highlighting |
+		( m_labelVisible * (int)GraphLayer::Overlay );
```
In 6eff5b5 I've switched this to a constant, which fixes hovered nodules.  We could add a new dirty mechanism for layers changing, but I we decided that there probably isn't much benefit currently.

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1098,29 +1098,47 @@ void ViewportGadget::renderInternal( Gadget::Layer filterLayer ) const
+
+			if( renderItem.style )
+			{
+				renderItem.style->bind();
```
Addressed in a360ed0

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1098,29 +1098,47 @@ void ViewportGadget::renderInternal( Gadget::Layer filterLayer ) const
+	{
+		// Actually overriding the style per Gadget is quite bad for performance, we're hoping this branch will
+		// never trigger, and style will be left at nullptr
+		style = gadget->getStyle();
```
Addressed in a360ed0

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -1096,7 +1098,7 @@ void ViewportGadget::renderInternal( Gadget::Layer filterLayer ) const
 			if( selector )
 			{
-				selector->loadName( renderItem.gadget->m_glName );
+				selector->loadName( i );
```
While trying to figure out why I couldn't make a test that showed this to be a problem, I realized that there's no reason to put something in RenderItems list if its layerMask is 0.  Fixing this meant that failing to increment by 1 would show up as a bug, and reduced time in "Moving Node ( Zoomed in )" test case that bottlenecks on adding render items from 51 milliseconds to 40 milliseconds.   The fix is in a396eb8, and the new test is in 0225e7c

danieldresser-ie commented on src/GafferUI/ViewportGadget.cpp:
```diff
@@ -943,13 +943,14 @@ std::vector< Gadget* > ViewportGadget::gadgetsAt( const Imath::Box2f &rasterRegi
+
+		// Hmmm, should I store the Gadget as non-const in m_renderItems for this purpose?  I guess the
+		// const cast is fine
+		gadgets.push_back( const_cast<Gadget*>( m_renderItems[ it.name ].gadget ) );
```
Removed the unnecessary comment while adding a new comment in a395eb8 ( I'm assuming this will all get squashed before merging anyway ).

danieldresser-ie commented:
Took longer than I was hoping getting the little stuff cleaned up today.  I've kept some things as little fix commits in a way that hopefully makes the reviewing easier, but this PR will get easier to work with once I squash it.

Of the bugs you've discussed, the nodule hovering one has been fixed, but I haven't yet added the new API that we're talking about for addressing render bounds different from layout bounds.  I've put up a sketch of what the header could look like here: https://github.com/danieldresser-ie/gaffer/commit/7950745433944aed4555353aa0c4e1eecbcf56c5

Does that look right to you?  You didn't really like the way I'm currently using a default bound to mean no bound for purposes of layout, but an infinite bound for the purposes of rendering - now that we're separating the two, I guess I can get rid of this, and force people to explicitly return Box3f::makeInfinite if they want an infinite bound.  I'm not sure what the default behaviour for the render bound should be ... I could make it behave the same as the layout bound by default, returning a union of the children, but I'm not sure how much sense this makes as a default ... we do have some Gadgets that implement a doRenderLayer that renders something based on the size of the bounds of their children, but I'm not sure this is a general rule?

I guess in a sense the most conservative would be to make the `bound( BoundType::Render )` return empty by default ( matching the empty default doRenderLayer ), and force anyone who implements doRenderLayer to also implement both layerMask and bound?  What do you think?

danieldresser-ie commented:
Still in the middle of working out the last few things, but I've now uploaded the implementations behind the new API additions in the last 2 commits, and it seems to be mostly working.

danieldresser-ie commented:
I think this is all pretty well good now.  Only bit that seems sketchy to me was trying to figure out how to bind BoundType in 194ef1d

I wasn't immediately sure how to get at the ImageGadget from the View to flip off useGPU, so I just added a MessageHandler to match what ImageGadget currently does.

johnhaddon commented:
Thanks Daniel! I've rebased on master and pushed a few extra commits for your consideration. Everything up to 84e7a5c is exactly as before. With the exception of 6fdf754 which I amended directly, all my other changes are in new commits.

> Only bit that seems sketchy to me was trying to figure out how to bind BoundType in 194ef1d

I took a shot at this in 84ec759, and it seems to work fine. Better than binding under a different name I think.

> I wasn't immediately sure how to get at the ImageGadget from the View to flip off useGPU, so I just added a MessageHandler to match what ImageGadget currently does.

It's available trivially as a plug on the view - amended in 6fdf754.

Hopefully the other commits are uncontentious - I think we discussed 44de8c7 already and the rest is mostly tidying. One thing definitely worth looking at is the implementation of `transformedBound()` though - I switched it back to use the layout bound, since I couldn't imagine any reason for wanting the render bound. This has made me wonder if we should hide the `BoundType` stuff from the public API entirely though - `RenderBound` is only really for the consumption of ViewportGadget, because nobody else has any access to the rendering functions. So I'm wondering if it should look like this :

```
public :
     Box3f bound() const; // Non-virtual, returns `bound( BoundType::Layout )`.
protected :
     enum BoundType;
     virtual Box3f bound( BoundType bound ); // Or `doBound()`?
```

or even this :

```
public :
    virtual Box3f bound() const; // As before, returns layout bound
protected :
    virtual Box3f renderBound() const;
```

I'm not sure I feel strongly enough about this to deal with it now, but it does definitely seem to me that the render bound concept is purely for ViewportGadget's benefit, and therefore shouldn't be public?

johnhaddon commented:
> It's available trivially as a plug on the view - amended in 6fdf754.

And of course, that didn't work :) I've brought back your version in ed33a36, adding the missing `import IECore` that was making the tests fail last time.

danieldresser-ie commented:
I've squashed in all the little fixes.

> One thing definitely worth looking at is the implementation of transformedBound() though - I switched it back to use the layout bound, since I couldn't imagine any reason for wanting the render bound.

Oops, that was definitely a mistake.

> So I'm wondering if it should look like this ( with renderBound() )

Maybe this was overkill when we're trying to get this wrapped up, but I actually find this dramatically clearer, so I've had a go at it that way ( I've cached a version locally with everything squashed and without doing this change if you want to go back to that ).  I can't really think of a downside though.  I guess there a few places where need to explicitly do `renderBound() { return bound() }` - but I feel like this is actually pretty clear, since it reflects a doRenderLayer method that uses bound().

johnhaddon commented on src/GafferUI/BackdropNodeGadget.cpp:
```diff
@@ -341,6 +341,13 @@ unsigned BackdropNodeGadget::layerMask() const
+Imath::Box3f BackdropNodeGadget::renderBound() const
+{
+	// If it's important to render text that extends outside the backdrop, we should include the header bound
+	// here - but we don't have access to style here?
```
We can access that with `this->style()`.

danieldresser-ie replied:
Getting this right ended up being pretty annoying, so for the moment I've just made the comment more accurate.

johnhaddon commented on Changes.md:
```diff
@@ -15,10 +17,17 @@ Breaking Changes
 - ViewportGadget : Deprecated old `gadgetsAt()` signature. Please use the new form instead.
 - ArnoldOptions : Removed support for the `ai:ignore_motion_blur` option. Turn off the `sampleMotion` option using a StandardOptions node instead.
 
+Improvements
```
Improvements sections should always be before API and Breaking Changes, so we're directing attention where we want it.

johnhaddon commented on Changes.md:
```diff
@@ -15,10 +17,17 @@ Breaking Changes
+Improvements
+------------
+
+- Better UI performance when navigating large node graphs
```
- GraphEditor : Improved drawing performance for large node graphs, \<insert some relevant numbers\>.

johnhaddon commented on python/GafferSceneUITest/CropWindowToolTest.py:
```diff
@@ -117,6 +118,8 @@ def testImageViewStatus( self ) :
 		script["image"] = GafferImage.ImageReader()
 
 		view = GafferUI.View.create( script["image"]["out"] )
+		# Avoid errors on CI machines without hardware graphics.
+		view["lutGPU"].setValue( False )
```
This turned out to be completely useless, so we can remove it.

johnhaddon commented on python/GafferSceneUITest/CropWindowToolTest.py:
```diff
@@ -152,9 +155,16 @@ def testImageViewStatus( self ) :
 
-		self.waitForIdle( 1000 )
-		self.assertEqual( tool.status(), "Info: Editing <b>options.options.renderCropWindow.value</b>" )
+		with IECore.CapturingMessageHandler() as mh :
```
Looks like we still have the test failure on CI - I think we're going to have to broaden the scope of this `with` block, potentially all the way up to where we make the window.

johnhaddon reviewed (changes requested):
> Maybe this was overkill when we're trying to get this wrapped up, but I actually find this dramatically clearer, so I've had a go at it that way

Sounds good - I was kindof hoping you'd prefer that way. Couple of tiny tweaks still needed (see inline comments) - could you just squash those in so we're ready to merge?

danieldresser-ie commented:
Addressed the last few comments.
